hoot-core/src/main/cpp/hoot/core/io/ApiDbReader.cpp

```diff
@@ -402,19 +404,110 @@ void ApiDbReader::_readByBounds(OsmMapPtr map, const Envelope& bounds)
+        //  Iterate any new nodes that are members of relations that need to be queried
+        newNodes = newNodes.subtract(nodeIds);
+        if (newNodes.size() > 0)
+        {
+          boost::shared_ptr<QSqlQuery> nodeItr = _getDatabase()->selectElementsByElementIdList(newNodes, TableType::Node);
+          while (nodeItr->next())
+          {
+            const QSqlQuery resultIterator = *nodeItr;
+            NodePtr node = _resultToNode(resultIterator, *map);
+            LOG_VART(node);
+            map->addElement(node);
+            boundedNodeCount++;
+            const long nodeId = resultIterator.value(0).toLongLong();
+            LOG_VART(ElementId(ElementType::Node, nodeId));
+            nodeIds.insert( QString::number(nodeId));
+          }
+        }
+        //  Iterate any new ways that are members of relations that need to be queried
+        newWays = newWays.subtract(wayIds);
+        if (newWays.size() > 0)
+        {
+          QSet<QString> additionalNodeIds;
+          LOG_DEBUG("Retrieving ways by way ID...");
+          boost::shared_ptr<QSqlQuery> wayItr =
+            _getDatabase()->selectElementsByElementIdList(newWays, TableType::Way);
+          while (wayItr->next())
+          {
+            WayPtr way = _resultToWay(*wayItr, *map);
+            map->addElement(way);
+            LOG_VART(way);
+            boundedWayCount++;
+          }
+
+          LOG_DEBUG("Retrieving way node IDs referenced by the selected ways...");
+          boost::shared_ptr<QSqlQuery> additionalWayNodeIdItr =
+            _getDatabase()->selectWayNodeIdsByWayIds(newWays);
+          while (additionalWayNodeIdItr->next())
+          {
+            const long nodeId = (*additionalWayNodeIdItr).value(0).toLongLong();
+            LOG_VART(ElementId(ElementType::Node, nodeId));
+            additionalNodeIds.insert(QString::number(nodeId));
+          }
+
+          //subtract nodeIds from additionalWayNodeIds so no dupes get added
+          LOG_VARD(nodeIds.size());
+          LOG_VARD(additionalNodeIds.size());
+          additionalNodeIds = additionalNodeIds.subtract(nodeIds);
+          LOG_VARD(additionalNodeIds.size());
+
+          if (additionalNodeIds.size() > 0)
+          {
+            LOG_DEBUG(
+              "Retrieving nodes falling outside of the query bounds but belonging to a selected way...");
+            boost::shared_ptr<QSqlQuery> additionalWayNodeItr =
+              _getDatabase()->selectElementsByElementIdList(additionalNodeIds, TableType::Node);
+            while (additionalWayNodeItr->next())
+            {
+              NodePtr node = _resultToNode(*additionalWayNodeItr, *map);
+              map->addElement(node);
+              LOG_VART(node);
+              boundedNodeCount++;
+            }
+          }
         }
+        //  Get the set of new relations found minus anything that has already been completed
+        relationIds = newRelations.subtract(completedRelationIds);
```

Comment:
Finally put any new relations found back into the `relationsId` container that is the basis of the `while` loop above minus any relation that is already in the map.